### ISSUE TITLE
5.0.0 -> 5.0.1 mysql foreign_keys is crazy slow

### ISSUE THREAD

**Issue by grosser** [rails/rails#27579]
```
Benchmark.realtime { ActiveRecord::Base.connection.foreign_keys('users') }
=> 26.654776551004034
```

makes our `rake db:schema:dump` take 5 minutes

can try out with https://github.com/zendesk/samson

I'd think https://github.com/rails/rails/pull/25307 is responsible

FYI flamegraph

<img width="913" alt="screen shot 2017-01-04 at 7 54 09 pm" src="https://cloud.githubusercontent.com/assets/11367/21668530/a25596ae-d2b7-11e6-9075-f210beb1cd5d.png">

**Comment by grosser** [comment]
@kamipo

**Comment by rafaelfranca** [comment]
@grosser are you able to run a quick explain query with the query before and after and see the time difference? I can do that but since you already have the setup to reproduce it would help to narrow down the reason for this performance change.

**Comment by grosser** [comment]
```
User.connection.select_all <<-SQL
          EXPLAIN SELECT fk.referenced_table_name AS 'to_table',
                 fk.referenced_column_name AS 'primary_key',
                 fk.column_name AS 'column',
                 fk.constraint_name AS 'name',
                 rc.update_rule AS 'on_update',
                 rc.delete_rule AS 'on_delete'
          FROM information_schema.key_column_usage fk
          JOIN information_schema.referential_constraints rc
          USING (constraint_schema, constraint_name)
          WHERE fk.referenced_column_name IS NOT NULL
            AND fk.table_schema = 'users'
            AND fk.table_name = 'users'
SQL

#<ActiveRecord::Result:0x007fe0e4b85700 @columns=["id", "select_type", "table", "type", "possible_keys", "key", "key_len", "ref", "rows", "Extra"], @rows=[[1, "SIMPLE", "fk", "ALL", nil, "TABLE_SCHEMA,TABLE_NAME", nil, nil, nil, "Using where; Open_full_table; Scanned 0 databases"], [1, "SIMPLE", "rc", "ALL", nil, nil, nil, nil, nil, "Using where; Open_full_table; Scanned all databases; Using join buffer"]], @hash_rows=nil, @column_types={}>
```

I'd guess the 'Scanned all databases' is the cause since I got plenty databases on my machine ...

**Comment by grosser** [comment]
old sql:

```
User.connection.select_all <<-SQL
          EXPLAIN SELECT fk.referenced_table_name as 'to_table'
                 ,fk.referenced_column_name as 'primary_key'
                 ,fk.column_name as 'column'
                 ,fk.constraint_name as 'name'
          FROM information_schema.key_column_usage fk
          WHERE fk.referenced_column_name is not null
            AND fk.table_schema = 'users'
            AND fk.table_name = 'users'
SQL

#<ActiveRecord::Result:0x007fe0e3183708 @columns=["id", "select_type", "table", "type", "possible_keys", "key", "key_len", "ref", "rows", "Extra"], @rows=[[1, "SIMPLE", "fk", "ALL", nil, "TABLE_SCHEMA,TABLE_NAME", nil, nil, nil, "Using where; Open_full_table; Scanned 0 databases"]], @hash_rows=nil, @column_types={}>
```

**Comment by rafaelfranca** [comment]
Awesome! That seems to be the reason indeed.

**Comment by jeremy** [comment]
@grosser Try adding `AND rc.table_name='users'` to the query. That should limit it to a single db.

Ref https://dev.mysql.com/doc/refman/5.7/en/information-schema-optimization.html

**Comment by grosser** [comment]
Yep!
```
Benchmark.realtime { ActiveRecord::Base.connection.foreign_keys('users') }
=> 0.16480079499888234
```

and still has the correct results ;)

**Comment by kamipo** [comment]
Sorry for the performance regression :cry:
I also confirmed the fix and created the PR #27580.

```
> EXPLAIN SELECT fk.referenced_table_name AS 'to_table', fk.referenced_column_name AS 'primary_key', fk.column_name AS 'column', fk.constraint_name AS 'name', rc.update_rule AS 'on_update', rc.delete_rule AS 'on_delete' FROM information_schema.key_column_usage fk JOIN information_schema.referential_constraints rc USING (constraint_schema, constraint_name) WHERE fk.referenced_column_name IS NOT NULL AND fk.table_schema = 'activerecord_unittest' AND fk.table_name = 'fk_test_has_pk' AND rc.table_name = 'fk_test_has_pk'\G
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: fk
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: TABLE_SCHEMA,TABLE_NAME
      key_len: NULL
          ref: NULL
         rows: NULL
     filtered: NULL
        Extra: Using where; Open_full_table; Scanned 0 databases
*************************** 2. row ***************************
           id: 1
  select_type: SIMPLE
        table: rc
   partitions: NULL
         type: ALL
possible_keys: NULL
          key: TABLE_NAME
      key_len: NULL
          ref: NULL
         rows: NULL
     filtered: NULL
        Extra: Using where; Open_full_table; Scanned 1 database; Using join buffer (Block Nested Loop)
2 rows in set, 1 warning (0.00 sec)
```
